app/gui/view/graph-editor/src/component/node/input/widget.rs

```diff
@@ -293,136 +370,44 @@ impl Configuration {
-        let is_expected_arg = kind.is_expected_argument();
-
-        let usage_type = info.usage_type.as_ref().map(|t| t.as_str());
-        let decl_type = kind.tp().map(|t| t.as_str());
-        let decl_or_usage = decl_type.or(usage_type);
-
-        let first_decl_is_vector = || {
-            decl_type
-                .map_or(false, |t| t.trim_start_matches('(').starts_with(list_editor::VECTOR_TYPE))
-        };
-        let type_may_be_vector = || {
-            decl_type.map_or(false, |t| t.contains(list_editor::VECTOR_TYPE))
-                || usage_type.map_or(false, |t| t.contains(list_editor::VECTOR_TYPE))
-        };
-        let allows_list = allow(F::ListEditor)
-            && info.connection.is_none()
-            && (looks_like_vector || (is_expected_arg && type_may_be_vector()));
-        let prefer_list = allows_list && first_decl_is_vector();
-        let tags = kind.tag_values().filter(|tags| !tags.is_empty());
-        let first_tag = tags.and_then(|t| t.first());
-
-        let mut config =
-            match (kind, tags) {
-                (Kind::Argument(_) | Kind::InsertionPoint(_), Some(tags))
-                    if (allow(F::SingleChoice) || prefer_list) =>
-                    Self::static_dropdown(kind.name().as_ref().map(Into::into), tags)
-                        .into_list_item_if(prefer_list, decl_or_usage, first_tag),
-
-                (Kind::Root | Kind::Argument(_), _) if allows_list =>
-                    Self::list_editor(None, decl_or_usage, first_tag),
-
-                (Kind::InsertionPoint(p), _)
-                    if p.kind.is_expected_argument() && (allow(F::Label) || allows_list) =>
-                    Self::always(label::Config::default()).into_list_item_if(
-                        allows_list,
-                        decl_or_usage,
-                        first_tag,
-                    ),
-
-                _ if allow(F::Hierarchy) && has_children => Self::always(hierarchy::Config),
-
-                (Kind::Token | Kind::Operation, _) if allow(F::Label) =>
-                    Self::inert(label::Config::default()),
-
-                (Kind::InsertionPoint(_), _) if allow(F::InsertionPoint) =>
-                    Self::inert(insertion_point::Config),
-
-                _ => {
-                    // Option of last resort, label is allowed in this case. Skip assert.
-                    return Self::always(label::Config::default());
+    fn automatic(ctx: &ConfigContext, disallowed: DynKindFlags) -> Self {
+        let allowed = !disallowed;
+        let mut best_match = None;
+        for kind in allowed {
+            let score = kind.match_node(ctx);
+            let current_score = best_match.map(|(_, score)| score).unwrap_or(Score::NotMatched);
+            if score > current_score {
+                best_match = Some((kind, score));
+                if score == Score::Perfect {
+                    break;
                 }
-            };
+            }
+        }
```

Comment:
I tried using iterators here, but I wasn't able to find anything nice here. It is important to me here to ensure that the iteration stops after first `Score::Perfect` is found. The closest solution is a chain of `take_while` and `max_by_key`, but that would stop the iteration one element too early (`take_while` would not pass the matched element through itself). I could write similar logic by using `try_fold` and storing extra flag externally, but I don't think it would be simpler than it is right now.

Reply:
You don't need a flag. You can check if current_score is already Perfect before checking of next element score.
```rust
    allowed.try_fold(None, |best_match, score| {
        let current_score = best_match.map(|(_, score)| score).unwrap_or(Score::NotMatched);
        (current_score != Score::Perfect).then(|| {
             let score = ....
        }
    }
```
But I'm not sure if it's actually cleaner, so we can stick to your version.